# Data analysis on IMDB scores of actors in Netflix movies/shows

## Goal

In this project, we aim to evaluate and clean the Netflix movies/shows rating dataset and to retrieve the actors with highest rating in each genre.

## Introduction

We have a dataset that includes various information of all Netflix movies/shows available in the United States from year 1945 to 2022. The dataset is represented by two tables `titles.csv` and `credits.csv`.

`titles.csv` describes information of Netflix movies/shows including titles, genres, ratings, whereas `credits.csv` has information of directors and actors that have participated in the movies/shows, including their names.

For each column of `titles.csv`:
- `id`: unique id of the movie/show.
- `title`: movie/show title.
- `type`: type of the show (movie or show).
- `description`: short description of the movie/show.
- `release_year`: the year that the movie/show was released.
- `age_certification`: age certificates for movies/shows.
- `runtime`: runtime of the movie/show.
- `genres`: genre of the movie/show.
- `production_countries`: the countries that the movie/show was produced.
- `seasons`: seasons of the show (not applicable for movies).
- `imdb_id`: IMDB id related to the movie/show.
- `imdb_score`: rating from IMDB.
- `imdb_votes`: number of votes from IMDB.
- `tmdb_popularity`: popularity of the movie/show from TMDB.
- `tmdb_score`: rating from TMDB.

For each column of `credits.csv`:
- `person_ID`: unique id of the director/actor.
- `id`: movie/show id that the person has participated in.
- `name`: name of the person.
- `character`: character name that the person has played in a movie/show.
- `role`: role of the person in a movie/show (directors or actors).

## Data Parsing

Use Pandas to read csv files and convert them to DataFrame. The `original_titles` table will contain the data from "title.csv", and `original_credits` will contain the data from "credits.csv".

In [1]:
import pandas as pd

In [2]:
original_titles = pd.read_csv("titles.csv")
original_credits = pd.read_csv("credits.csv")

Let's preview a few records from these tables.

In [3]:
original_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [4]:
original_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

## Data evaluation and cleansing

Here we are going to evaluate the data in `original_titles` and `original_credits` in terms of structure and validity. Then we will perform data cleansing if needed.

We make copies of these tables and assign them `cleaned_titles` and `cleaned_credits` for better readability.

In [5]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

### Evaluate the data structure

First we are going to check `cleaned_titles`.

In [6]:
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   
5   ts22164         Monty Python's Flying Circus   SHOW   
6   tm70993                        Life of Brian  MOVIE   
7   tm14873                          Dirty Harry  MOVIE   
8  tm119281                     Bonnie and Clyde  MOVIE   
9   tm98978                      The Blue Lagoon  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   
5  A British sketch comedy series with the shows ...          1969   
6  Brian Cohen is an average young Jewish man, bu...          1979   
7  When a madman dubbed 'Scorpio' terrorizes San ...          1971   
8  In the 1930s, bored waitress Bonnie Parker fal...          1967   
9  Two small children and a ship's cook survive a...          1980   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   
5             TV-14       30                       ['comedy', 'european']   
6                 R       94                                   ['comedy']   
7                 R      102              ['thriller', 'action', 'crime']   
8                 R      110                 ['crime', 'drama', 'action']   
9                 R      104               ['romance', 'action', 'drama']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   
5               ['GB']      4.0  tt0063929         8.8     73424.0   
6               ['GB']      NaN  tt0079470         8.0    395024.0   
7               ['US']      NaN  tt0066999         7.7    155051.0   
8               ['US']      NaN  tt0061418         7.7    112048.0   
9               ['US']      NaN  tt0080453         5.8     69844.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600  
5           17.617       8.306  
6           17.770       7.800  
7           12.817       7.500  
8           15.687       7.500  
9           50.324       6.156

From the first 10 records in the `cleaned_titles`, we observe that there are multiples values in the column `genres` and `production_countries`. In order to better analyze the data we will have to split the values to ensure one single value per column.

Firstly, let's deal with `genres`. We will look at its value type.

In [7]:
cleaned_titles['genres'][1]

"['drama', 'crime']"

Notice that each value in this column is a string instead of an actual list of strings indicating the genres. Therefore we will have to convert it to a list.

In [8]:
cleaned_titles['genres'] = cleaned_titles['genres'].apply(lambda s: eval(s))
cleaned_titles['genres'][1]

['drama', 'crime']

Then we will use `explode` method in DataFrame to split the values.

In [9]:
cleaned_titles = cleaned_titles.explode("genres")
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

Problem solved. Now let's look at `production_countries`.

In [10]:
cleaned_titles['production_countries'][1]

1    ['US']
1    ['US']
Name: production_countries, dtype: object

We notice the same issue as in `genres`. Let's convert them into a list of strings.

In [11]:
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].apply(lambda s: eval(s))
cleaned_titles['production_countries'][1]

1    [US]
1    [US]
Name: production_countries, dtype: object

Let's use `explode` again.

In [12]:
cleaned_titles = cleaned_titles.explode('production_countries')
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   
2                 R      109       thriller                   US      NaN   
2                 R      109       european                   US      NaN   
3                PG       91        fantasy                   GB      NaN   
3                PG       91         action                   GB      NaN   
3                PG       91         comedy                   GB      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

The structure of `cleaned_titles` looks good now. Let's move on to `cleaned_credits`.

In [13]:
cleaned_credits.head(10)

person_id       id             name                      character   role
0       3748  tm84618   Robert De Niro                  Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster                  Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                            Tom  ACTOR
3       3739  tm84618    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                          Betsy  ACTOR
5      32267  tm84618      Peter Boyle                         Wizard  ACTOR
6     519612  tm84618   Leonard Harris      Senator Charles Palantine  ACTOR
7      29068  tm84618   Diahnne Abbott                Concession Girl  ACTOR
8     519613  tm84618      Gino Ardito             Policeman at Rally  ACTOR
9       3308  tm84618  Martin Scorsese  Passenger Watching Silhouette  ACTOR

This looks perfect in terms of structure. We don't need to modify anything here.

### Evaluate the data validity

Let's look at `cleaned_titles`.

In [14]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64

There are 17818 entries in this table and we've observed that some columns like `title` and `description` have null values. We will have to evaluate them to see if we should fill in the missing data.

Also notice that `release_year` should be represented as date instead of int64.

In [15]:
cleaned_titles["release_year"] = pd.to_datetime(cleaned_titles["release_year"], format='%Y')
cleaned_titles["release_year"]

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

Let's also look at `cleaned_credits`.

In [16]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


There are 77801 entries in this table and we've observed that `character` has null values.

Also notice that `person_id` should be represented as string instead of int64.

In [17]:
cleaned_credits["person_id"] = cleaned_credits["person_id"].astype("str")
cleaned_credits["person_id"]

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

### Deal with null/missing data

In table `cleaned_titles`, we see that `title`, `description`, `age_certification`, `genres`, `production_countries`, `seasons`, `imdb_id`, `imdb_score`, `tmdb_popularity`, `tmdb_score`, `imdb_votes`, `tmdb_popularity`, `tmdb_score` have null values.

Realize that the purpose of this project is to find out actors from high IMDB rating movies and shows. Null values in `title`, `description`, `age_certification`, `production_countries`, `seasons`, `imdb_id`, `tmdb_popularity`, `tmdb_score`, `imdb_votes`, `tmdb_popularity`, `tmdb_score` should not bother us and we will keep them as is.

`imdb_score` and `genres` are closely related to what we are trying to analyze, so let's deep dive into those values.

First, let's look at `imdb_score`.

In [18]:
cleaned_titles.query("imdb_score.isnull()")

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

Since these IMDB scores are missing, we can't analyze the corresponding movies or shows. We will have to delete these entries.

In [19]:
cleaned_titles = cleaned_titles.dropna(subset=["imdb_score"])
cleaned_titles["imdb_score"].isnull().sum()

0

Now let's look at `genres`.

In [20]:
cleaned_titles.query("genres.isnull()")

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

We want to know great actors from different genres of movies/shows, so let's delete these entries with null values.

In [21]:
cleaned_titles = cleaned_titles.dropna(subset=["genres"])
cleaned_titles["genres"].isnull().sum()

0

In table `cleaned_credits`, we see that `character` has null values.

In [22]:
cleaned_credits.query("character.isnull()")

person_id         id              name character      role
36         3308    tm84618   Martin Scorsese       NaN  DIRECTOR
59        17727   tm154986      John Boorman       NaN  DIRECTOR
106       11475   tm127384       Terry Jones       NaN  DIRECTOR
107       11473   tm127384     Terry Gilliam       NaN  DIRECTOR
162        1063   tm120801    Robert Aldrich       NaN  DIRECTOR
...         ...        ...               ...       ...       ...
77776   2363022  tm1097142  Mohamed El-Arkan       NaN     ACTOR
77777   1827884  tm1097142     Mohamed Bakir       NaN  DIRECTOR
77783    678884  tm1014599      Segun Arinze       NaN     ACTOR
77789   1962840  tm1014599     Seyi Babatope       NaN  DIRECTOR
77800   1950416  tm1059008    Julian Gaviria       NaN  DIRECTOR

[9772 rows x 5 columns]

We only care about actor names but not the character names. Hence we don't have to deal with these entries.

### Deal with duplicate data

The table `cleaned_titles` should not have two or more entries with all same values. Let's check that.

In [23]:
cleaned_titles.duplicated().sum()

0

There are no duplicate entries, so we are good.

In [24]:
cleaned_credits.duplicated().sum()

0

Same check on `cleaned_credits`. There are no duplicates.

### Deal with inconsistent data

In [25]:
cleaned_titles['genres'].value_counts()

drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: genres, dtype: int64

Each value in `genres` refers to a unique genre of movies/shows. There is no such a case where different values refer to a same genre.

Sometimes, we might see inconsistent data in countries names. For example, both "US" and "United States" refer to the same country. We will check `production_countries` in table `cleaned_titles`.

In [30]:
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

US         5648
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK            9
MT      

Notice that there is a entry with value `Lebanon`. It actually refers to `LB` which is already included in the table. We are replacing `Lebanon` to `LB` in this case.

In [31]:
cleaned_titles["production_countries"] = cleaned_titles["production_countries"].replace({"Lebanon": "LB"})

In [32]:
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

US    5648
IN    1610
GB    1068
JP    1046
FR     720
KR     637
ES     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
AT      28
SA      28
IS      28
LU      27
NZ      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
UY      15
MA      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
MT       8
VN       8
LT       7
IR       7
CD       7
SU       7
TN       7
SN       6
AL       6
KE       6
GH       6
IQ       5
MU       5
CY       5
KN       4
GR       4
IO       4
SY       4
TZ       4
MC       4
GL       3
CM       3
AO       3
BS       3
HR       3
BD       3
PY       3
DZ       3

Let's check `role` in table `cleaned_credits`.

In [33]:
cleaned_credits['role'].value_counts()

ACTOR       73251
DIRECTOR     4550
Name: role, dtype: int64

In [34]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  object
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: object(5)
memory usage: 3.0+ MB


There are only two possible values in `role`: `ACTOR` or `DIRECTOR`. In this case we may convert the data type to `category` to save some memory.

In [35]:
cleaned_credits["role"] = cleaned_credits["role"].astype("category")
cleaned_credits["role"]

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

In [37]:
cleaned_titles.describe()

runtime      seasons    imdb_score    imdb_votes  tmdb_popularity  \
count  16970.000000  5954.000000  16970.000000  1.694100e+04     16842.000000   
mean      80.912552     2.455492      6.514207  3.281655e+04        29.396307   
std       39.596172     2.869428      1.131095  1.141492e+05        93.178235   
min        0.000000     1.000000      1.500000  5.000000e+00         0.600000   
25%       45.000000     1.000000      5.800000  7.800000e+02         4.070000   
50%       90.000000     2.000000      6.600000  3.508000e+03        10.195000   
75%      107.000000     3.000000      7.300000  1.697800e+04        23.639000   
max      225.000000    42.000000      9.500000  2.294231e+06      2274.044000   

         tmdb_score  
count  16515.000000  
mean       6.846933  
std        1.078831  
min        1.000000  
25%        6.200000  
50%        6.900000  
75%        7.500000  
max       10.000000

It looks like there are no invalid data like negative values.

## Retrieve the data that we want

In [38]:
cleaned_titles

id                               title   type  \
1       tm84618                         Taxi Driver  MOVIE   
1       tm84618                         Taxi Driver  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
...         ...                                 ...    ...   
5846   tm898842                         C/O Kaadhal  MOVIE   
5847  tm1059008                             Lokillo  MOVIE   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   

                                            description release_year  \
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
...                                                 ...          ...   
5846  A heart warming film that explores the concept...   2021-01-01   
5847  A controversial TV host and comedian who has b...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   

     age_certification  runtime     genres production_countries  seasons  \
1                    R      114      drama                   US      NaN   
1                    R      114      crime                   US      NaN   
2                    R      109      drama                   US      NaN   
2                    R      109     action                   US      NaN   
2                    R      109   thriller                   US      NaN   
...                ...      ...        ...                  ...      ...   
5846               NaN      134      drama                  NaN      NaN   
5847               NaN       90     comedy                   CO      NaN   
5849               NaN        7     family                  NaN      1.0   
5849               NaN        7  animation                  NaN      1.0   
5849               NaN        7     comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5846  tt11803618         7.7       348.0              NaN         NaN  
5847  tt14585902         3.8        68.0           26.005       6.300  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[16970 rows x 15 columns]

In [39]:
cleaned_credits

person_id         id                name                character  \
0          3748    tm84618      Robert De Niro            Travis Bickle   
1         14658    tm84618        Jodie Foster            Iris Steensma   
2          7064    tm84618       Albert Brooks                      Tom   
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4         48933    tm84618     Cybill Shepherd                    Betsy   
...         ...        ...                 ...                      ...   
77796    736339  tm1059008    Adelaida Buscato                María Paz   
77797    399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798    373198  tm1059008         Inés Prieto                    Fanny   
77799    378132  tm1059008        Isabel Gaona                   Cacica   
77800   1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

To analyze the best actors in each genre, we want to join the tables `cleaned_titles` and `cleaned_credits` by column `id` as it refers to movies/shows in both tables.

In [40]:
credits_with_titles = pd.merge(cleaned_credits, cleaned_titles, on="id", how="inner")

In [41]:
credits_with_titles.head()

person_id       id            name      character   role        title  \
0      3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
1      3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
2     14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
3     14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
4      7064  tm84618   Albert Brooks            Tom  ACTOR  Taxi Driver   

    type                                        description release_year  \
0  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  crime                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  crime                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0         8.2    808582.0           40.965       8.179  
1         8.2    808582.0           40.965       8.179  
2         8.2    808582.0           40.965       8.179  
3         8.2    808582.0           40.965       8.179  
4         8.2    808582.0           40.965       8.179

Since we only care about actors, let's filter out the unwanted entries.

In [42]:
actor_with_titles = credits_with_titles.query('role == "ACTOR"')

Use `groupby` to group `genres` and `person_id`. Then, we use `mean()` to calculate the average IMDB ratings for an actor in each genre.

In [43]:
groupby_genres_and_person_id = actor_with_titles.groupby(["genres", "person_id"])

In [44]:
imdb_score_groupby_genres_and_person_id = groupby_genres_and_person_id["imdb_score"].mean()
imdb_score_groupby_genres_and_person_id

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

We may reset the indice to get a better formatted DataFrame.

In [45]:
imdb_score_groupby_genres_and_person_id_df = imdb_score_groupby_genres_and_person_id.reset_index()
imdb_score_groupby_genres_and_person_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

Now we have the ratings for every actor in each genre. We also want to find the actor with highest rating in each genre. To do so we first need to find the maximum value in each genre and join the table `imdb_score_groupby_genres_and_person_id_df`.

In [46]:
genres_max_scores = imdb_score_groupby_genres_and_person_id_df.groupby("genres")["imdb_score"].max()
genres_max_scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

In [47]:
genres_max_score_with_person_id = pd.merge(imdb_score_groupby_genres_and_person_id_df, genres_max_scores, on=["genres", "imdb_score"])
genres_max_score_with_person_id

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

Now we have what we need. Let's also print out their names.

In [49]:
actor_id_with_names = cleaned_credits[['person_id', 'name']].drop_duplicates()
genres_max_score_with_actor_name = pd.merge(genres_max_score_with_person_id, actor_id_with_names, on="person_id")
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0       action     12790         9.3         Olivia Hack
1        scifi     12790         9.3         Olivia Hack
2       action      1303         9.3       Jessie Flower
3    animation      1303         9.3       Jessie Flower
4       family      1303         9.3       Jessie Flower
..         ...       ...         ...                 ...
131        war    826547         8.8         Yuto Uemura
132    western     22311         8.9     Koichi Yamadera
133    western     28166         8.9  Megumi Hayashibara
134    western     28180         8.9      Unsho Ishizuka
135    western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

We may sort the table by genres for a better view.

In [51]:
genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values("genres").reset_index().drop("index", axis=1)
genres_max_score_with_actor_name.head(20)

genres person_id  imdb_score             name
0      action     12790         9.3      Olivia Hack
1      action    336830         9.3  André Sogliuzzo
2      action     21033         9.3       Zach Tyler
3      action     86591         9.3    Cricket Leigh
4      action      1303         9.3    Jessie Flower
5   animation     28024         9.3      Dante Basco
6   animation      1303         9.3    Jessie Flower
7   animation     86591         9.3    Cricket Leigh
8   animation     21033         9.3       Zach Tyler
9   animation    336830         9.3  André Sogliuzzo
10     comedy    140194         9.2     Ryu Jun-yeol
11     comedy     44199         9.2    Kim Sung-kyun
12     comedy   1296454         9.2     Lee Se-young
13     comedy     71076         9.2        Ra Mi-ran
14     comedy    747462         9.2         Kim Seol
15     comedy    140196         9.2     Ahn Jae-hong
16     comedy    101292         9.2    Ryu Hye-young
17     comedy     38441         9.2      Park Bo-gum
18     comedy    103374         9.2       Lee Hye-ri
19      crime     20552         9.5     Betsy Brandt

Now we have reached our goal! Let's store the data as a file.

In [52]:
genres_max_score_with_actor_name.to_csv("genres_max_score_with_actor_name.csv", index=False)